pkg/storegateway/indexheader/lazy_binary_reader.go

```diff
@@ -223,7 +223,7 @@ func (r *LazyBinaryReader) LabelNames() ([]string, error) {
 
 // load ensures the underlying binary index-header reader has been successfully loaded. Returns
 // an error on failure. This function MUST be called with the read lock already acquired.
-func (r *LazyBinaryReader) load() (returnErr error) {
+func (r *LazyBinaryReader) load() error {
```

Comment:
I've split `load` function into `load` and `load1`. First one must be called with read-lock, as before. `load1` is called without any locks, and will itself acquire (and release) write-lock.